File: frontend/src/services/pricing.js
import { resolveApiUrl } from '../config/api';

// Unsplash bundled local images for reliable offline fallback
const sareeImg = require('../../assets/images/products/banarasi-saree.jpg');
const elephantImg = require('../../assets/images/products/wood-elephant.jpg');
const potImg = require('../../assets/images/products/terracotta-pot.jpg');

const FALLBACK_BENCHMARKS = [
  {
    id: 'c1',
    title: 'Pure Katan Silk Handloom Banarasi Saree',
    imageUrl: sareeImg,
    price: 6800,
    source_label: 'Varanasi Weavers Cooperative Benchmark'
  },
  {
    id: 'c2',
    title: 'Hand-Carved Sheesham Wood Royal Elephant',
    imageUrl: elephantImg,
    price: 3800,
    source_label: 'Saharanpur Wood Carvers Benchmark'
  },
  {
    id: 'c3',
    title: 'Hand-Thrown Terracotta Decorative Vase',
    imageUrl: potImg,
    price: 2400,
    source_label: 'Khurja Pottery Cluster Benchmark'
  }
];

/**
 * Suggests fair dynamic pricing based on:
 * - Material cost
 * - Craft category margins (published Ministry of Textiles & FTFI guidelines)
 * - GI certified status (+25-40% premium)
 *
 * Calls real FastAPI backend: POST /api/pricing/suggest
 */
export async function suggestPricing({ craftType = 'Handloom Weaving', materialCost = 2200, isGiMatch = false, productCategory = null }) {
  const url = resolveApiUrl('/api/pricing/suggest');

  try {
    const response = await fetch(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Accept': 'application/json',
      },
      body: JSON.stringify({
        craft_type: craftType,
        material_cost: parseFloat(materialCost) || 2200,
        product_category: productCategory,
        gi_match_status: Boolean(isGiMatch),
      }),
    });

    if (!response.ok) {
      const errText = await response.text();
      throw new Error(`Pricing API returned HTTP ${response.status}: ${errText}`);
    }

    const data = await response.json();
    return {
      success: true,
      minPrice: data.min_price,
      recommendedPrice: data.recommended_price,
      maxPrice: data.max_price,
      materialCost: data.material_cost,
      markupMultiplierRange: data.markup_multiplier_range,
      giMultiplierApplied: data.gi_multiplier_applied,
      giPremiumPct: data.gi_premium_pct,
      craftCategoryMatched: data.craft_category_matched,
      citation: data.citation,
      method: data.method,
      disclaimer: data.disclaimer,
      marketComparables: data.market_comparables || [],
      isLiveBackend: true,
    };
  } catch (error) {
    console.warn('[pricing.js] Real pricing API request failed, applying transparent local heuristic fallback:', error.message);

    // Honest fallback using the exact same rules-based formula
    const cost = Math.max(10, parseFloat(materialCost) || 2200);
    const baseMin = 2.4;
    const baseRec = 2.9;
    const baseMax = 3.4;
    const giMult = isGiMatch ? 1.30 : 1.0;

    const rawMin = cost * baseMin * (isGiMatch ? 1.25 : 1.0);
    const rawRec = cost * baseRec * giMult;
    const rawMax = cost * baseMax * (isGiMatch ? 1.40 : 1.0);

    const minPrice = Math.ceil(rawMin / 100) * 100;
    const recommendedPrice = Math.ceil(rawRec / 100) * 100;
    const maxPrice = Math.ceil(rawMax / 100) * 100;

    return {
      success: true,
      minPrice,
      recommendedPrice,
      maxPrice,
      materialCost: cost,
      markupMultiplierRange: { min: baseMin, recommended: baseRec, max: baseMax },
      giMultiplierApplied: giMult,
      giPremiumPct: isGiMatch ? 30 : 0,
      craftCategoryMatched: craftType || 'Handloom & Textiles',
      citation: 'Ministry of Textiles / FTFI Artisan Costing Norms (Local Heuristic Fallback)',
      method: 'rules_based_heuristic_v1_local_fallback',
      disclaimer: 'Calculated using local rules-based heuristic fallback due to network timeout.',
      marketComparables: FALLBACK_BENCHMARKS,
      isLiveBackend: false,
      errorNotice: error.message,
    };
  }
}

/**
 * Retrieves market comparison benchmarks for a craft category.
 * Calls real FastAPI backend: GET /api/pricing/benchmarks
 */
export async function getMarketComparison(craftTypeId) {
  const url = resolveApiUrl(`/api/pricing/benchmarks?craft_type=${encodeURIComponent(craftTypeId || 'handloom')}`);

  try {
    const response = await fetch(url, {
      method: 'GET',
      headers: { Accept: 'application/json' },
    });

    if (response.ok) {
      const items = await response.json();
      if (Array.isArray(items) && items.length > 0) {
        // Map to format expected by SmartPricing UI (with bundled image fallbacks)
        return items.map((item, idx) => ({
          id: item.id,
          title: item.title,
          price: item.price,
          imageUrl: idx === 0 ? sareeImg : (idx === 1 ? elephantImg : potImg),
          sourceLabel: item.source_label,
        }));
      }
    }
  } catch (error) {
    console.warn('[pricing.js] Could not fetch market benchmarks from backend, using bundled reference items:', error.message);
  }

  return FALLBACK_BENCHMARKS;
}

File: backend/app/routers/pricing.py
from typing import List, Optional, Dict, Any
from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel, Field

from app.services.pricing_service import pricing_service

router = APIRouter(prefix="/api/pricing", tags=["Dynamic Pricing Engine"])

class PricingSuggestRequest(BaseModel):
    craft_type: Optional[str] = Field("Handloom Weaving", description="Artisan craft category or product type")
    material_cost: float = Field(..., description="Artisan-entered raw material cost in INR (₹)")
    product_category: Optional[str] = Field(None, description="Optional secondary craft classification")
    gi_match_status: Optional[bool] = Field(False, description="Whether product has verified GI-certified status")

class MarketComparableItem(BaseModel):
    id: str
    title: str
    price: int
    source_label: str

class PricingSuggestResponse(BaseModel):
    min_price: int
    recommended_price: int
    max_price: int
    material_cost: float
    markup_multiplier_range: Dict[str, float]
    base_markup_range: Dict[str, float]
    gi_multiplier_applied: float
    gi_premium_pct: int
    craft_category_matched: str
    citation: str
    method: str = "rules_based_heuristic_v1"
    disclaimer: str
    market_comparables: List[MarketComparableItem]

@router.post(
    "/suggest",
    response_model=PricingSuggestResponse,
    summary="Suggest Fair Dynamic Pricing (Material Cost + Published Margins + GI Premium)"
)
def suggest_pricing(req: PricingSuggestRequest):
    """
    Dynamic Pricing Suggestion:
    - Base multiplier off artisan-entered material cost based on published handicraft costing guidelines.
    - Adjustment for GI-certified status (25-40% empirical realization premium, using 30% midpoint).
    - Returns suggested range (min/recommended/max).
    - Method explicitly labeled 'rules_based_heuristic_v1' with full transparency disclaimer.
    - Curated static reference table of realistic market-comparables returned alongside.
    """
    if req.material_cost <= 0:
        raise HTTPException(status_code=400, detail="Material cost must be greater than zero.")

    try:
        result = pricing_service.calculate_suggestion(
            craft_type=req.craft_type,
            material_cost=req.material_cost,
            product_category=req.product_category,
            gi_match_status=bool(req.gi_match_status)
        )
        return PricingSuggestResponse(**result)
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Pricing calculation failed: {str(e)}")

@router.get(
    "/benchmarks",
    response_model=List[MarketComparableItem],
    summary="Get Static Curated Market Comparables for Craft Category"
)
def get_benchmarks(
    craft_type: Optional[str] = Query(None, description="Craft type (e.g., Handloom, Pottery, Brass, Wood, Embroidery)")
):
    """
    Returns curated reference benchmark items for market comparison.
    Explicitly documented as static reference data, not live-scraped feeds.
    """
    items = pricing_service.get_benchmarks_for_category(craft_type)
    return [MarketComparableItem(**it) for it in items]
